File: rfcs/bench-0104/harness/js/knucleotide.js
// k-nucleotide, the JavaScript leg of RFC-0104 M2.
//
// The specification is examples/knucleotide.vyrn. Same tables, same ordering
// (count descending, ties by fragment ascending), byte-identical output.
//
//   $ node knucleotide.js < ../../fasta-1000.expected
//
// Structural differences from the Vyrn program:
//
//   * a k-mer key is `seq.slice(i, i + k)` on a JavaScript string. The Vyrn
//     program builds an Array<UInt8> window and pays one `stringFromBytes` per
//     position — a UTF-8 validation over ASCII. This leg does not do that.
//   * one `Array.prototype.sort` with a comparator replaces Vyrn's two passes (a
//     hand-written insertion sort by fragment, then a stable `sortBy` on the
//     negated count), because Vyrn's sortBy takes an Int64 key and no comparator
//     crosses the API. Same total order, one pass instead of two.
//   * stdin is read once rather than a line at a time, so this leg holds the
//     whole input; the Vyrn program holds only the THREE sequence.
//   * toUpperCase() is Unicode-aware where std/strings toUpper is ASCII. The
//     input is ASCII, so the two agree here.

'use strict';
const fs = require('fs');

let pending = '';
function emit(s) {
    pending += s;
    if (pending.length >= 1 << 20) flush();
}
function flush() {
    if (pending.length === 0) return;
    const buf = Buffer.from(pending, 'latin1');
    pending = '';
    let off = 0;
    while (off < buf.length) {
        try {
            off += fs.writeSync(1, buf, off, buf.length - off);
        } catch (e) {
            if (e.code !== 'EAGAIN') throw e;
        }
    }
}

/// The five fragments the game asks for by name, after the two frequency tables.
function namedFragments() {
    return ['GGT', 'GGTA', 'GGTATT', 'GGTATTTTAATT', 'GGTATTTTAATTTATAGT'];
}

/// The THREE sequence from FASTA on stdin: uppercased, with the newlines and the
/// other two sequences left out.
function thirdSequence() {
    const parts = [];
    let inThird = false;
    for (const l of fs.readFileSync(0, 'latin1').split('\n')) {
        if (l.startsWith('>')) {
            inThird = l.startsWith('>THREE');
        } else if (inThird) {
            parts.push(l.toUpperCase());
        }
    }
    return parts.join('');
}

/// Every window of width `k`, counted.
function countKmers(seq, k) {
    const m = new Map();
    let i = 0;
    while (i + k <= seq.length) {
        const key = seq.slice(i, i + k);
        const seen = m.get(key);
        m.set(key, seen === undefined ? 1 : seen + 1);
        i = i + 1;
    }
    return m;
}

/// `x` at three decimal places — the game asks for three.
function fixed3(x) {
    const scaled = Math.trunc(x * 1000.0 + 0.5);
    const whole = Math.trunc(scaled / 1000);
    const frac = String(scaled % 1000).padStart(3, '0');
    return `${whole}.${frac}`;
}

/// Count descending, ties by fragment ascending.
function ranked(m) {
    const es = [];
    for (const [frag, count] of m) {
        es.push({ frag: frag, count: count });
    }
    es.sort((a, b) => {
        if (a.count !== b.count) return b.count - a.count;
        return a.frag < b.frag ? -1 : a.frag > b.frag ? 1 : 0;
    });
    return es;
}

/// One frequency table: every fragment of width `k` as a percentage of the
/// windows there are, then a blank line.
function report(seq, k) {
    const m = countKmers(seq, k);
    const total = seq.length - k + 1;
    for (const e of ranked(m)) {
        emit(`${e.frag} ${fixed3(100.0 * e.count / total)}\n`);
    }
    emit('\n');
}

/// The count of one named fragment.
///
/// It builds the whole table for that width and looks the fragment up, rather
/// than scanning for the one string — because the table IS the benchmark.
function countOf(seq, frag) {
    const m = countKmers(seq, frag.length);
    const c = m.get(frag);
    return c === undefined ? 0 : c;
}

const seq = thirdSequence();
report(seq, 1);
report(seq, 2);
for (const frag of namedFragments()) {
    emit(`${countOf(seq, frag)}\t${frag}\n`);
}
flush();

File: rfcs/bench-0104/harness/rust/knucleotide.rs
// k-nucleotide, from the Computer Language Benchmarks Game.
//
// The Rust leg of RFC-0104 M2. It computes what `examples/knucleotide.vyrn`
// computes and prints the same bytes.
//
//   $ rustc -C opt-level=3 -o knucleotide.exe knucleotide.rs
//   $ ./knucleotide < ../../fasta-1000.expected
//
// Safe Rust, std only: `std::collections::HashMap` with its default hasher, and
// no external hash crate. There is no N: the input is FASTA on stdin.
//
// The keys are borrowed `&[u8]` windows of the sequence, which is what a Rust
// program would naturally write — `seq.windows(k)` allocates nothing. The Vyrn
// program allocates a `String` per window because `Map` is keyed by `String`.

use std::collections::HashMap;
use std::io::{self, BufRead, BufWriter, Write};

/// The five fragments the game asks for by name, after the two frequency tables.
const NAMED_FRAGMENTS: [&str; 5] = [
    "GGT",
    "GGTA",
    "GGTATT",
    "GGTATTTTAATT",
    "GGTATTTTAATTTATAGT",
];

/// The THREE sequence from FASTA on stdin: uppercased, with the newlines and the
/// other two sequences left out.
fn third_sequence(input: &mut impl BufRead) -> Vec<u8> {
    let mut out = Vec::new();
    let mut in_third = false;
    for line in input.lines() {
        let l = line.unwrap();
        if l.starts_with('>') {
            in_third = l.starts_with(">THREE");
        } else if in_third {
            out.extend(l.bytes().map(|b| b.to_ascii_uppercase()));
        }
    }
    out
}

/// Every window of width `k`, counted.
fn count_kmers(seq: &[u8], k: usize) -> HashMap<&[u8], i64> {
    let mut m: HashMap<&[u8], i64> = HashMap::new();
    for w in seq.windows(k) {
        *m.entry(w).or_insert(0) += 1;
    }
    m
}

/// `x` at three decimal places — the game asks for three.
fn fixed3(x: f64) -> String {
    let scaled = (x * 1000.0 + 0.5) as i64;
    format!("{}.{:03}", scaled / 1000, scaled % 1000)
}

/// Count descending, ties by fragment ascending.
fn ranked(m: HashMap<&[u8], i64>) -> Vec<(&[u8], i64)> {
    let mut es: Vec<(&[u8], i64)> = m.into_iter().collect();
    es.sort_by(|a, b| b.1.cmp(&a.1).then_with(|| a.0.cmp(b.0)));
    es
}

/// One frequency table: every fragment of width `k` as a percentage of the
/// windows there are, then a blank line.
fn report(out: &mut impl Write, seq: &[u8], k: usize) {
    let m = count_kmers(seq, k);
    let total = seq.len() as i64 - k as i64 + 1;
    for (frag, count) in ranked(m) {
        writeln!(
            out,
            "{} {}",
            std::str::from_utf8(frag).unwrap(),
            fixed3(100.0 * count as f64 / total as f64)
        )
        .unwrap();
    }
    writeln!(out).unwrap();
}

/// The count of one named fragment. It builds the whole table for that width and
/// looks the fragment up, because the table IS the benchmark.
fn count_of(seq: &[u8], frag: &str) -> i64 {
    let m = count_kmers(seq, frag.len());
    *m.get(frag.as_bytes()).unwrap_or(&0)
}

fn main() {
    let stdin = io::stdin();
    let mut input = stdin.lock();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());

    let seq = third_sequence(&mut input);
    report(&mut out, &seq, 1);
    report(&mut out, &seq, 2);
    for frag in NAMED_FRAGMENTS {
        writeln!(out, "{}\t{}", count_of(&seq, frag), frag).unwrap();
    }
}
